klippy/extras/exclude_object.py

```diff
@@ -0,0 +1,263 @@
+
+        self.last_position[:] = newpos
+        self.last_position_extruded[:] = self.last_position
+        self.max_position_extruded = max(self.max_position_extruded, newpos[3])
+
+        # These next few conditionals handle the moves immediately after leaving
+        # and excluded object.  The toolhead is at the end of the last printed
+        # object and the gcode is at the end of the last excluded object.
+        #
+        # Ideally, there will be Z and E moves right away to adjust any offsets
+        # before moving away from the last position.  Any remaining corrections
+        # will be made on the firs XY move.
+        if (offset[0] != 0 or offset[1] != 0) and \
+            (newpos[0] != self.last_position_excluded[0] or \
+            newpos[1] != self.last_position_excluded[1]):
+            offset[0] = 0
+            offset[1] = 0
+            offset[2] = 0
+            offset[3] += self.extruder_adj
+            self.extruder_adj = 0
+
+        if offset[2] != 0 and newpos[2] != self.last_position_excluded[2]:
+            offset[2] = 0
+
+        if self.extruder_adj != 0 and \
+            newpos[3] != self.last_position_excluded[3]:
+            offset[3] += self.extruder_adj
+            self.extruder_adj = 0
+
+        tx_pos = newpos[:]
+        for i in range(4):
+            tx_pos[i] = newpos[i] - offset[i]
+        self.next_transform.move(tx_pos, speed)
+
+    def _ignore_move(self, newpos, speed):
+        offset = self.extrusion_offsets[self.extruder_idx]
+        for i in range(3):
+            offset[i] = newpos[i] - self.last_position_extruded[i]
+        offset[3] = offset[3] + newpos[3] - self.last_position[3]
+        self.last_position[:] = newpos
+        self.last_position_excluded[:] =self.last_position
+        self.max_position_excluded = max(self.max_position_excluded, newpos[3])
+
+    def _move_into_excluded_region(self, newpos, speed):
+        self.in_excluded_region = True
+        self._ignore_move(newpos, speed)
+
+    def _move_from_excluded_region(self, newpos, speed):
+        self.in_excluded_region = False
+        offset = self.extrusion_offsets[self.extruder_idx]
+
+        # This adjustment value is used to compensate for any retraction
+        # differences between the last object printed and excluded one.
+        self.extruder_adj = self.max_position_excluded \
+            - self.last_position_excluded[3] \
+            - (self.max_position_extruded - self.last_position_extruded[3])
+        self._normal_move(newpos, speed)
+
+    def _test_in_excluded_region(self):
+        # Inside cancelled object
+        return self.current_object in self.excluded_objects
+
+    def get_status(self, eventtime=None):
+        status = {
+            "objects": list(self.objects.values()),
+            "excluded_objects": list(self.excluded_objects),
+            "current_object": self.current_object
+        }
+        return status
```

Comment:
Why is the `objects` field a list instead of the more natural dictionary?  That is, wouldn't it be more useful to return `"objects": self.objects`?  I'm a little concerned about the overhead of translating the object dict to a list on every get_status invocation (many UIs subscribe to everything, so internally we have to build the status every 250ms and check for changes).

Reply:
Even though this code uses a map, it seemed that for a status, a simple list made more sense.  It's been a long time since I changed this code, but I remember having issues with status not updating when passing those values without wrapping them in the `list()` calls.

Reply:
Right - if the dict is exported from `get_status()` then it needs to be treated as immutable internally.  See the recent conversation at #5222 .  Basically, make copies of the dict on every change to the dict instead of on every `get_status()` call.